### DISCUSSION TITLE
convert .csv to Tensor

### DISCUSSION THREAD

**Discussion by ShayanRiyaz** [onnx/onnx#2745] [Q&A]
Hi, I haven't found the relevant documentation for this issue, therefore, I am asking the question over here.

I have a model trained using tensorflow.keras and the model reads CSV files as an input. I have attached the whole process in the zip file along with the model converted to a .onnx file. I'm currently using Unity for a project and I am trying to give my trained .onnx model input of a csv file (in my case it will be a 20,9) file. However, I am unable to understand how to convert the csv file into tensor input in the project.

note: The .csv file is created within the unity project
```
28.75, 120.02, 44.02, 102.95, 14.57, 128.26, 21.05, 104.83, 6.01
28.68, 120.01, 44.00, 102.94, 14.48, 128.35, 20.77, 104.62, 5.86
28.61, 120.19, 44.00, 103.27, 14.32, 128.69, 20.35, 104.89, 5.61
8.42, 120.21, 44.08, 103.44, 14.66, 128.94, 20.81, 105.04, 5.91
28.47, 120.29, 44.06, 103.59, 14.43, 128.78, 20.43, 104.62, 5.88
28.46, 120.09, 44.01, 103.57, 14.59, 128.70, 20.68, 104.69, 6.08
28.33, 120.10, 43.95, 103.81, 14.44, 128.94, 20.71, 104.33, 5.91
28.27, 120.03, 43.89, 103.94, 14.44, 128.94, 20.80, 104.79, 5.84
28.24, 119.79, 43.72, 103.67, 14.28, 128.40, 21.03, 105.20, 5.85
28.16, 119.82, 43.71, 103.69, 14.13, 128.58, 20.72, 104.91, 5.94
28.17, 119.88, 43.83, 103.73, 14.36, 128.62, 20.66, 104.88, 5.79
28.06, 119.87, 43.83, 103.69, 14.49, 128.53, 20.69, 104.83, 5.84
28.11, 119.78, 43.82, 103.60, 14.32, 128.41, 20.57, 104.70, 5.72
27.91, 119.71, 43.83, 103.59, 14.41, 128.48, 20.49, 104.57, 5.79
27.81, 119.75, 43.82, 103.63, 14.43, 128.69, 20.64, 104.94, 5.91
27.73, 119.66, 43.67, 103.67, 14.21, 128.50, 20.50, 104.67, 5.79
27.97, 119.65, 43.87, 103.63, 14.70, 128.41, 20.93, 105.15, 5.69
27.78, 119.67, 43.72, 103.78, 14.24, 128.72, 20.61, 104.96, 5.79
27.91, 119.66, 43.91, 103.67, 14.58, 128.46, 20.46, 104.81, 5.77
27.81, 119.58, 43.82, 103.65, 14.56, 128.67, 20.68, 104.82, 5.65

```

I found this C# online but it is for a color image
```
public static Tensor TransformInput(Color32[] pic, int width, int height)
{
    float[] floatValues = new float[width * height * 3];

    for (int i = 0; i < pic.Length; ++i)
    {
        var color = pic[i];

        floatValues[i * 3 + 0] = (color.r - IMAGE_MEAN) / IMAGE_STD;
        floatValues[i * 3 + 1] = (color.g - IMAGE_MEAN) / IMAGE_STD;
        floatValues[i * 3 + 2] = (color.b - IMAGE_MEAN) / IMAGE_STD;
    }

    return new Tensor(1, height, width, 3, floatValues);
}
```
If anyone can help me transform the input or guide me to the relevant documentation, that would be great.

Thank you.

[random_files_tensorflow.zip](https://github.com/onnx/onnx/files/4533790/random_files_tensorflow.zip)

**Comment by TMVector** [comment]
Hi @ShayanRiyaz, it's not entirely clear what you want to do. It looks like you want to convert a CSV to a [Tensor](https://www.nuget.org/packages/System.Numerics.Tensors) so you can call ONNX Runtime from C#?

If so, you should refer to the [ONNX Runtime C# API docs](https://github.com/microsoft/onnxruntime/blob/master/docs/CSharp_API.md), which includes a link to a sample. I didn't find the docs very helpful -- it looks like the Tensors library is not really mainstream yet -- so you might want to see the source too https://github.com/dotnet/runtime/tree/master/src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors.

I think the gist of it is:
```csharp
using Microsoft.ML.OnnxRuntime.Tensors;
//..
float[,] values = ReadYour2dCsvHere("file/path.csv");
return values.ToTensor();
```
